src/ht.c

```diff
@@ -1,239 +1,238 @@
- * the rights to use, copy, modify, merge, publish, distribute, sublicense,
- * and/or sell copies of the Software, and to permit persons to whom the
- * Software is furnished to do so, subject to the following conditions:
- *
- * The above copyright notice and this permission notice (including the next
- * paragraph) shall be included in all copies or substantial portions of the
- * Software.
- *
- * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
- * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
- * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
- * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
- * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
- * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
- * IN THE SOFTWARE.
- *
- * Except as contained in this notice, the names of the authors
- * or their institutions shall not be used in advertising or
- * otherwise to promote the sale, use or other dealings in this
- * Software without prior written authorization from the
- * authors.
- *
- * Authors:
- *    Eric Anholt <eric@anholt.net>
- *    Keith Packard <keithp@keithp.com>
- * Integration in r2 core api and hackit up
- *    pancake <nopcode.org>
- */
-
 #include "ht.h"
+#include "sdb.h"
+
+// Sizes of the ht.
+const int ht_primes_sizes[] = {
+3, 7, 11, 17, 23, 29, 37, 47, 59, 71, 89, 107, 131, 163, 197, 239, 293, 353, 431, 521, 631, 761, 919,
+1103, 1327, 1597, 1931, 2333, 2801, 3371, 4049, 4861, 5839, 7013, 8419, 10103, 12143, 14591,
+17519, 21023, 25229, 30293, 36353, 43627, 52361, 62851, 75431, 90523, 108631, 130363, 156437,
+187751, 225307, 270371, 324449, 389357, 467237, 560689, 672827, 807403, 968897, 1162687, 1395263,
+1674319, 2009191, 2411033, 2893249, 3471899, 4166287, 4999559, 5999471, 7199369};
+
+
+// Create a new hashtable and return a pointer to it.
+// size - number of buckets in the hashtable
+// hashfunction - the function that does the hashing, must not be null.
+// comparator - the function to check if values are equal, if NULL, just checks == (for storing ints).
+// keydup - function to duplicate to key (eg strdup), if NULL just does =.
+// valdup - same as keydup, but for values
+// pair_free - function for freeing a keyvaluepair - if NULL just does free.
+// calcsize - functino to calculate the size of a value. if NULL, just stores 0.
+SdbHash* internal_ht_new(ut32 size, HashFunction hashfunction, ListComparator comparator, DupKey keydup, DupValue valdup, HtKvFreeFunc pair_free, CalcSize calcsize) {
+	SdbHash* ht;
+	ut32 i;
+	ht = calloc (1, sizeof (*ht));
+	if (ht) {
+		ht->size = size;
+		ht->count = 0;
+		ht->prime_idx = 0;
+		ht->load_factor = 1;
+		ht->hashfn = hashfunction;
+		ht->cmp = comparator;
+		ht->dupkey = keydup;
+		ht->dupvalue = valdup;
+		ht->table = calloc (ht->size, sizeof (SdbList*));
+		ht->calcsize = calcsize;
+		if (ht->table) {
```

Comment:
do not initialize all items here, we should initialize the list only when it is necessary, otherwise NULL is better than r_list_empty()